### PR TITLE
Added heading and file icons

### PR DESCRIPTION
**Pull request by ema2159**
The following features were added (inspired in Centaur Emacs configuration):

- Added icons for the default headings and added a defcustom for enabling and disabling them.
- Added icons for the files and added a defcustom for enabling and disabling them.
- Added info about the packages loaded and the init time and added a defcustom for enabling and disabling it.

It was heavily inspired in the Centaur Emacs custom version of Dashboard. I also included a couple of enhancements like adding bullet icons for the org agenda or detecting when a file is remote (tramp) and add a "remote" icon, without establishing a connection (which Centaur Emacs didn't do) which makes it much more efficient.
 
I tried to keep it consistent with the repo coding style and I tried to make it the most readable possible (though elisp is a language that I am not very familiar with and I've not used it much besides from configuring my emacs).

It currently looks like this:
screenshot

### CODE REVIEW

**Comment by ema2159** [comment]
Could you help me to pass the linting checks? I fixed all the indicated errors but it still gives me the incorrect indentation error and doesn't tell me why. Please tell me what is it so that I can fix it.

**Comment by JesusMtnez** [comment]
Thank you @ema2159 for the hard work! I'll try to take a look at it during this weekend. :clap:

**Comment by JesusMtnez** [comment]
I find the problem with the lint. You need to update `.dir-locals.el`:
```
((emacs-lisp-mode . ((indent-tabs-mode . nil)
                     (fill-column . 100)
                     (elisp-lint-indent-specs . ((when-let . 1))))))
```
That would set `when-let` to only required one level of indentation, which is the one being applied in the macro by default.

**Comment by JesusMtnez** [comment]
@ema2159 One little thing. I really like the changes you propose. I want them to be included in the next `1.4.0` release. But one little thing. Would you mind to split the changes in two PRs? One PR including all `all-the-icons` integration (it could be this one) and the other one with the new init information.

File: dashboard-widgets.el [hunk 1]
```diff
@@ -294,6 +303,85 @@ WIDGET-PARAMS are passed to the \"widget-create\" function."
...
+             ((string-equal heading "Registers:")
+              (all-the-icons-octicon "database" :height 1.2 :v-adjust 0.0 :face 'dashboard-heading))
+             ((string-equal heading "Projects:")
+              (all-the-icons-octicon "rocket" :height 1.2 :v-adjust 0.0 :face 'dashboard-heading))))
+    (insert " ")))
+(advice-add #'dashboard-insert-heading :before #'dashboard-insert-heading-icon)
+
+;;
+;; Add file icons
+;; MUST redefine the sections because of the macro `dashboard-insert-section-list'
```
Comment (JesusMtnez) [inline dashboard-widgets.el]: This comment is not necessary

**Comment by ema2159** [comment]
> @ema2159 One little thing. I really like the changes you propose. I want them to be included in the next `1.4.0` release. But one little thing. Would you mind to split the changes in two PRs? One PR including all `all-the-icons` integration (it could be this one) and the other one with the new init information.

Yes sir, I'll do that as soon as I'm out of the office, probably between today's afternoon and tomorrow.
And just to be clear with the .dir-locals.el, I just need to create a .dir-locals.el in the same directory as the dashboard.el file and write `((emacs-lisp-mode . ((indent-tabs-mode . nil)
                     (fill-column . 100)
                     (elisp-lint-indent-specs . ((when-let . 1))))))`
in it?

File: dashboard-widgets.el (hunk 1 above)
Comment (ema2159) [reply dashboard-widgets.el]: Which one? or both of them?

Comment (JesusMtnez) [reply dashboard-widgets.el]: line 334

**Comment by JesusMtnez** [comment]
No rush on the changes :wink: 

There is a `.dir-locals.el` already in the repository with configuration for `emacs-lisp-mode`. You just need to add the element `(elisp-lint-indent-specs . ((when-let . 1))))` to the alist.

**Comment by ema2159** [comment]
Ok perfect. I also wanted to ask you what do you consider good defaults? I mean, I put all the defcustoms in nil thinking that maybe someone didn't had all-the-icons installed but I'm also thinking that people may miss that the icons option is available.
By the way, I'm pretty sure we both hablamos español a la perfección porque es nuestra lengua madre but we'll keep it in english for standard comunication jaja.

**Comment by JesusMtnez** [comment]
There were two errors failing in the CI. One was the indentation check that is fixed now. The other one is:
```elisp
In toplevel form:
dashboard-widgets.el:396:49:Error: ‘(path (car (last (split-string (abbreviate-file-name el) " - "))))’ is a malformed function
```

This happens when trying to byte-compile `dashboard-widgets.el`. Any idea why?

**Comment by ema2159** [comment]
> There were two errors failing in the CI. One was the indentation check that is fixed now. The other one is:
> 
> ```emacs-lisp
> In toplevel form:
> dashboard-widgets.el:396:49:Error: ‘(path (car (last (split-string (abbreviate-file-name el) " - "))))’ is a malformed function
> ```
> 
> This happens when trying to byte-compile `dashboard-widgets.el`. Any idea why?

Yes, I think Emacs 25 doesn't support the when-let macro. Changing that for a separated let, when expression seems to fix it

**Comment by ema2159** [comment]
Do you know exactly how do I have to include the all the icons package so that it doesn't complain about free variables or functions?

**Comment by JesusMtnez** [comment]
> Do you know exactly how do I have to include the all the icons package so that it doesn't complain about free variables or functions?

In `.emacs/dependencies.el`

**Comment by ema2159** [comment]
For some reason the master test failed to install page break lines and that made that test fail, but the others seem ok.

**Comment by JesusMtnez** [comment]
I was preparing a new PR to update emacs version in the CircleCI configuration. And I found the same problem. Let me work on that one later (WIP https://github.com/emacs-dashboard/emacs-dashboard/pull/124). Once that one gets merged, we should be good to go!

**Comment by ema2159** [comment]
Seems that we're done! I've separated the PR as you told me and everything seems to be working fine! If there's anything else I can help you with just let me know!

**Comment by JesusMtnez** [comment]
https://github.com/emacs-dashboard/emacs-dashboard/pull/124 is merged! The problem was a missing coding setting. Feel free to rebase this branch against master or merge master in this branch (as you prefer).

**Comment by ema2159** [comment]
Done, all the checks are passing now

**Comment by seagle0128** [comment]
Nice to see the features of Centaur Emacs are merged! Thank you!

**Comment by ema2159** [comment]
> Nice to see the features of Centaur Emacs are merged! Thank you!
> 
> One question: Did you test this scenario: If user doesn't install `all-the-icons`, but enable `dashboard-set-heading-icons` and/or `dashboard-set-file-icons`?
> 
> I also add some comments below.

No sir, I personally haven't, but I guess it probably just display ugly icons as it usually does when the icons aren't installed, that's why I defaulted the icon's defcustom to nil, but it shouldn't be a bigger problem than that because all the icons was added as a dependency. 
Also we're having this issue #128. Any idea why?
By the way, that's a killer configuration you've created.

**Comment by seagle0128** [comment]
@ema2159 I was not talking about icons, but about the `all-the-icons.el` package. If users didn't install this package, undefined `all-the-icons-xxx` will be called. I use `require 'all-the-icons nil t` because the package is installed in `Centaur Emacs` definitely.

Regarding #128 , I have to say it was as designed, since the icons are not widgets/buttons or the part of them, for displaying correctly. The buttons have different face properties which will break the properties of icons. That's why the `TAB` (next widget) does work but the cursor on icons doesn't.

**Comment by ema2159** [comment]
> @ema2159 I was not talking about icons, but about the `all-the-icons.el` package. If users didn't install this package, undefined `all-the-icons-xxx` will be called. I use `require 'all-the-icons nil t` because the package is installed in `Centaur Emacs` definitely.
> 
> Regarding #128 , I have to say it was as designed, since the icons are not widgets/buttons or the part of them, for displaying correctly. The buttons have different face properties which will break the properties of icons. That's why the `TAB` (next widget) does work but the cursor on icons doesn't.

Yes I understand, but now all the icons is a dependency so it's instalation is ensured with use package.

And regarding #128, I understand, so a work around would be that when the shorcut buttons are pressed the cursor fell directly into the text instead of the icon, do you know how to do this?

**Comment by seagle0128** [comment]
Okay. Glad you understand the risks.

Aha, and I have a great news, I think I found the solution for #123 . Should I submit a PR?

**Comment by JesusMtnez** [comment]
Of course, feel free to submit a PR and contribute! :smile:

**Comment by seagle0128** [comment]
Already committed. Please review and merge. Thanks!

**Pull request merged**